libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/VoiceInstructionLoader.java

```diff
@@ -0,0 +1,85 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import com.mapbox.api.directions.v5.models.LegStep;
+import com.mapbox.api.directions.v5.models.VoiceInstructions;
+import com.mapbox.api.speech.v1.MapboxSpeech;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class VoiceInstructionLoader {
+  private static VoiceInstructionLoader instance;
+  private MapboxSpeech.Builder mapboxSpeech;
+
+  private VoiceInstructionLoader() {
+
+  }
+
+  public static synchronized VoiceInstructionLoader getInstance() {
+    if (instance == null) {
+      instance = new VoiceInstructionLoader();
+    }
+
+    return instance;
+  }
+
+  public void initialize(MapboxSpeech.Builder builder) {
+    mapboxSpeech = builder;
+  }
+
+  public void getInstruction(String instruction, String textType, Callback<ResponseBody> callback) {
+    mapboxSpeech.instruction(instruction)
+      .textType(textType)
+      .build()
+      .enqueueCall(callback);
+  }
+
+  public void cacheInstruction(String instruction) {
+    mapboxSpeech.instruction(instruction).build().enqueueCall(new Callback<ResponseBody>() {
+      @Override
+      public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
+
+      }
+
+      @Override
+      public void onFailure(Call<ResponseBody> call, Throwable throwable) {
+
+      }
+    });
+  }
+
+  public void cacheInstructions(RouteProgress routeProgress, boolean first) {
```

Comment:
IMO `cacheInstructions` is getting too long.
What about [decomposing these conditionals](https://sourcemaking.com/refactoring/decompose-conditional) extracting blocks of code into private methods and make intention clearer?